internal/internal_task_pollers.go

```diff
@@ -235,13 +313,121 @@ func (wtp *workflowTaskPoller) ProcessTask(task interface{}) error {
-	wtp.metricsScope.Timer(metrics.DecisionEndToEndLatency).Record(time.Now().Sub(workflowTask.pollStartTime))
-	wtp.metricsScope.Counter(metrics.DecisionTaskCompletedCounter).Inc(1)
+	return nil
+}
 
+func newLocalActivityPoller(params workerExecutionParameters, laTunnel *localActivityTunnel) *localActivityTaskPoller {
+	handler := &localActivityTaskHandler{
+		userContext:  params.UserContext,
+		metricsScope: params.MetricsScope,
+		logger:       params.Logger,
+	}
+	return &localActivityTaskPoller{
+		handler:      handler,
+		metricsScope: params.MetricsScope,
+		logger:       params.Logger,
+		laTunnel:     laTunnel,
+	}
+}
+
+func (latp *localActivityTaskPoller) PollTask() (interface{}, error) {
+	return latp.laTunnel.getTask(), nil
+}
+
+func (latp *localActivityTaskPoller) ProcessTask(task interface{}) error {
+	result := latp.handler.executeLocalActivityTask(task.(*localActivityTask))
+	latp.laTunnel.deliverResult(result)
 	return nil
 }
 
+func (lath *localActivityTaskHandler) executeLocalActivityTask(task *localActivityTask) (result *localActivityResult) {
+	lath.metricsScope.Counter(metrics.LocalActivityTotalCounter).Inc(1)
+	activityType := getFunctionName(task.params.ActivityFn)
+	ae := activityExecutor{name: activityType, fn: task.params.ActivityFn}
+
+	rootCtx := lath.userContext
+	if rootCtx == nil {
+		rootCtx = context.Background()
+	}
+
+	ctx := context.WithValue(rootCtx, activityEnvContextKey, &activityEnvironment{
+		activityType:      ActivityType{Name: activityType},
+		activityID:        fmt.Sprintf("%v", task.activityID),
+		workflowExecution: task.params.WorkflowInfo.WorkflowExecution,
+		logger:            lath.logger,
+		metricsScope:      lath.metricsScope,
+		isLocalActivity:   true,
+	})
+
+	// panic handler
+	defer func() {
+		if p := recover(); p != nil {
+			topLine := fmt.Sprintf("local activity for %s [panic]:", activityType)
+			st := getStackTraceRaw(topLine, 7, 0)
+			lath.logger.Error("LocalActivity panic.",
+				zap.String("PanicError", fmt.Sprintf("%v", p)),
+				zap.String("PanicStack", st))
+			lath.metricsScope.Counter(metrics.LocalActivityPanicCounter).Inc(1)
+			panicErr := newPanicError(p, st)
+			result = &localActivityResult{
+				task:   task,
+				result: nil,
+				err:    panicErr,
+			}
+		}
+		if result.err != nil {
+			lath.metricsScope.Counter(metrics.LocalActivityFailedCounter).Inc(1)
+		}
+	}()
+
+	deadline := time.Now().Add(time.Duration(task.params.ScheduleToCloseTimeoutSeconds) * time.Second)
+	ctx, cancel := context.WithDeadline(ctx, deadline)
+	task.Lock()
+	if task.canceled {
+		task.Unlock()
+		return &localActivityResult{err: ErrCanceled, task: task}
+	}
+	task.cancelFunc = cancel
+	task.Unlock()
+
+	var laResult []byte
+	var err error
+	doneCh := make(chan struct{})
+	go func(ch chan struct{}) {
+		laStartTime := time.Now()
+		laResult, err = ae.ExecuteWithActualArgs(ctx, task.params.InputArgs)
```

Comment:
We are making an assumption that activity is going to cancel itself once context expires. I think it is reasonable assuption, but let's add a warning when activity actual execution time exceeded the deadline by some value.